### ISSUE TITLE
BUG: `randint` on Windows appears to require int32 not int

### ISSUE THREAD

**Issue by matthew-brett** [numpy/numpy#27498]
### Describe the issue:

The docstring of `np.random.randint` implies that the `high` parameter should be an `int`, but it appears, in fact, it requires an `np.intc`:

```python
[ins] In [2]: MAX_INT = np.iinfo(int).max

[ins] In [3]: MAX_INT
Out[3]: 9223372036854775807

[ins] In [4]: np.random.randint(MAX_INT)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[4], line 1
----> 1 np.random.randint(MAX_INT)

File numpy\\random\\mtrand.pyx:796, in numpy.random.mtrand.RandomState.randint()

File numpy\\random\\_bounded_integers.pyx:2881, in numpy.random._bounded_integers._rand_int32()

ValueError: high is out of bounds for int32
```

### Reproduce the code example:

```python
import numpy as np
MAX_INT = np.iinfo(int).max
np.random.randint(MAX_INT)
```

### Error message:

```shell
ValueError                                Traceback (most recent call last)
Cell In[4], line 1
----> 1 np.random.randint(MAX_INT)

File numpy\\random\\mtrand.pyx:796, in numpy.random.mtrand.RandomState.randint()

File numpy\\random\\_bounded_integers.pyx:2881, in numpy.random._bounded_integers._rand_int32()

ValueError: high is out of bounds for int32
```

### Python and NumPy Versions:

2.1.1
3.12.1 (tags/v3.12.1:2305ca5, Dec  7 2023, 22:03:25) [MSC v.1937 64 bit (AMD64)]

### Context for the issue:

See above.

**Comment by charris** [comment]
> numpy.random.mtrand.RandomState.randint()

RandomState is probably the problem, we no longer maintain it. Are you on windows? `int32` used to be the default integer on windows. Does the problem persist with the new random functions?

**Comment by seberg** [comment]
That is correct, the old randomstate still uses `long` throughout, while the new one uses `int64`,  so neither uses the "default integer" on NumPy 2.  (although with the new rng you can pass the dtype I think)

Use `np.dtype('long').max`, I suppose, unless you want to not use `long` (since its weird on windows).

**Comment by rkern** [comment]
This is a documentation issue. All of the `RandomState` methods that work with integers use C `long`s, which used to be what corresponded to `np.dtype(int)` before 2.0, so the correspondence was more apt then.

I'm mostly fine with continuing to use `int` in the parameter documentation (we really only mean "integer" not specifically exactly `np.dtype(int)`), but maybe putting a `Note` somewhere in the `RandomState` class docstring to talk about the platform-dependency. I'd be fine with having a reference to that note in each method that has such a parameter, but probably not the full text.

I'm also fine with letting `RandomState` wither on the vine, frankly.

**Comment by seberg** [comment]
It is already documented in the `dtype` argument (arguably, it affects the default as well of course):
```
    .. warning::
      This function defaults to the C-long dtype, which is 32bit on windows
      and otherwise 64bit on 64bit platforms (and 32bit on 32bit ones).
      Since NumPy 2.0, NumPy's default integer is 32bit on 32bit platforms
      and 64bit on 64bit platforms.  Which corresponds to `np.intp`.
      (`dtype=int` is not the same as in most NumPy functions.)
```

**Comment by matthew-brett** [comment]
Yes, that is confusing; it wasn't obvious to me that I needed to look at the dtype argument to know what numbers I could send to `high`.

**Comment by Harish3005-cloud** [comment]
Hi all, I would like to work on this documentation fix.

Based on the discussion above, I understand that `RandomState` methods rely on C-long types, which are limited to 32-bit on Windows. This causes unexpected `ValueErrors` when users pass large integers (like `MAX_INT`) to the `high` parameter.

**Plan:**
I will update the docstring for `randint` in `numpy/random/mtrand.pyx`. specifically adding a `.. note::` to the `high` parameter description to explicitly warn Windows users about the 32-bit limit (similar to the existing warning in `dtype`).

**Comment by mattip** [comment]
The change should be a few words, maybe not even a full `.. note:`: "... int32 on Windows, int64 elsewhere" not much more than that.
